oak_idl_gen_services/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! This crate allows compiling flatbuffer services to Rust in `build.rs` scripts.
+//!
+//! Also see the `oak_idl_gen_structs` crate.
+//!
+//! These crates are intentionally distinct so that `oak_idl_gen_services` can use
+//! `oak_idl_gen_structs` in order to compile the flatbuffers reflection schema at compile time.
+
+#![feature(path_file_prefix)]
+#![feature(iter_intersperse)]
+
+use anyhow::Context;
+use convert_case::Casing;
+use reflection_generated::reflection::{RPCCall, Service};
+use std::process::exit;
+
+mod reflection_generated {
+    #![allow(clippy::derivable_impls, clippy::needless_borrow)]
+    #![allow(dead_code, unused_imports)]
+
+    include!(concat!(env!("OUT_DIR"), "/reflection_generated.rs"));
+}
+
+/// Compile services from the provided flatbuffer file using the `flatc` binary installed on the
+/// system.
+///
+/// Services are generated targeting the invocation-based transport from the `oak_idl` crate (i.e.
+/// not gRPC).
+///
+/// Each service method in the `rpc_service` definition must have a unique `method_id` numeric
+/// attribute associated, which is used as part of the invocation serialization process. The
+/// attribute must be declared upfront in the same file.
+///
+/// ```text
+/// namespace Test;
+///
+/// attribute "method_id";
+///
+/// table Request {}
+/// table Response {}
+///
+/// rpc_service Name {
+///     Method(Request) : Response (method_id: 42);
+/// }
+/// ```
+///
+/// For a service called `Name`, this method generates the following objects:
+///
+/// - a struct named `NameClient`, exposing a method for each method defined in the service. This
+///   may be used to connect to underlying transport and to invoke methods on the corresponding
```

Comment:
Should we clarify this a bit perhaps? It seems we could say that _in tests_ it can be used to (directly) invoke methods on the corresponding `Server` object. In real use cases this will always be used to connect to an underlying transport channel.  :)

Reply:
thx! :)